README.md
# logmatic-docker
*Link to the [Logmatic.io documentation](http://doc.logmatic.io/docs/docker)*

Releases and tags

* `1.2`, `latest`: The official Logmatic.io image
* `1.0` (*deprecated*): NodeJS docker client, not compatible with the Logmatic.io integration
* `dev`: dev-build from the github repository

The Logmatic.io's container finds all your running containers' logs, events and stats and stream them to your platform.

# Use the image available on docker hub

This container as for unique option to pass the api key of your Logmatic.io's platform.

So to use the image available on docker hub simply run the following command:

```
docker run -dt --name logmatic.io -v /var/run/docker.sock:/var/run/docker.sock:ro logmatic/logmatic-docker:1.2 <your_api_key>
```

The mapping to the docker socket is really important as this is why we are able to identify all the running containers and follow their logs.

Nothing more to do.

# Options

Several options are allowed after the api key.

```
> usage: logmatic-docker [-h] [--no-ssl] [--no-logs] [--no-stats]
                 [--no-detailed-stats] [--no-events] [--namespace NS]
                 [--hostname HOSTNAME] [--port PORT] [--debug] [-i INTERVAL]
                 [--attr ATTRS] [--docker-version VER] [--skipByImage REGEX]
                 [--skipByName REGEX] [--matchByImage REGEX]
                 [--matchByName REGEX] [--matchByLabel LABEL]
                 LOGMATIC_API_KEY
  
  Send logs, events and stats to Logmatic.io
  
  positional arguments:
    LOGMATIC_API_KEY      The Logmatic.io API key
  
  optional arguments:
    -h, --help            show this help message and exit
    --no-logs             Disable the logs streams
    --no-stats            Disable the stats streams
    --no-detailed-stats   Disable stats streams
    --no-events           Disable the event stream
    --namespace NS        Default namespace
    --hostname HOSTNAME   Logmatic.io's hostname (default api.logmatic.io)
    --port PORT           Logmatic.io's port (default 10514)
    --debug               Enable debugging
    -i INTERVAL           Seconds between to stats report (default 30)
    --attr ATTRS          eg myattribute="my attribute"
    --docker-version VER  Force the Docker version to use
    --skipByImage REGEX   Skip container by image name
    --skipByName REGEX    Skip container by container name
    --matchByImage REGEX  Match container by image name
    --matchByName REGEX   Match container by container name
    --matchByLabel LABEL  Format either "key" or "key=value"
```


## Add extra attributes

You can add extra attributes to all the pushed entries by chaining the option "--attr" or "-a".

## Match / Skip by name, image or label

If you don't want all your containers to send log entries to Logmatic.io you can user the options
`--matchByLabel`, `--matchByImage`, `--matchByName`, `--skipByImage` or `--skipByName`.

However, use one inclusion/exclusion policy as these options cannot live together.


## Disable docker container stats

You can disable container stats for each container using `--no-stats` or `--no-detailed-stats` if you want to keep only the
a summary of the usage. You can also set the interval with `-i INTERVAL` (set to
30 seconds by default).


## Disable docker events

You can disable container events for each container using `--no-events`

## Namespace the docker attributes

By default all the attributes related to Docker such as the container id, the name, the image etc... are contained into the `docker` object (as illustrated below). You can then use this parameter to change this namespace. With an empty namespace `--namespace ""`: all these attributes are merged at the root level.

# What are the data types sent to Logmatic.io?

This container sends 3 types of data:

- the log data published by your containers
- the docker events
- the docker stats

## Log data

The log data is published as is associated with some meta-data that identifies the container.
An NGINX log for instance would look like this:

```json
{
    "message": "192.168.99.1 - - [15/Dec/2015:17:36:50 +0000] \"GET / HTTP/1.1\" 304 0 \"-\" \"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_9_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/47.0.2526.80 Safari/537.36\" \"-\"",
    "docker": {
      "image": "agileek/cpuset-test",
      "hostname": "f307d28f60f4",
      "daemon_name": "jarvis",
      "created": "2016-12-21T15:04:27.6401807Z",
      "name": "small_mahavira",
      "id": "f307d28f60f4695e90bd108b3a3bbb6c9fbd4e896187966515b9d18d104f1730",
      "short_id": "f307d28f60",
      "status": "running",
      "pid": 7931
    }
}
```

**NOTE about JSON logging:**

If the container logs in JSON, the message field contains a valid JSON object. Logmatic.io will then parse it by default and you'll get all your attributes properly created at the root level in the user interface.

## Docker events

Docker events tells you the activity on your docker machine: create, kill, commit, etc...

```json
{
    "message": "[Docker event] name:distracted_archimedes >> event:attach (image=ubuntu)",
    "docker": {
      "image": "ubuntu",
      "daemon_name": "jarvis",
      "name": "distracted_archimedes",
      "id": "65254830bfa3604f487d1603d509d45886f37f83412c4b18f809d148c5f60c4c",
      "event": "attach",
      "status": "attach"
    }
}
```

## Docker stats

Docker stats are all the metrics that matters by container. And there are a lot of them.

```json
{
    "message": "[Docker stats] name:small_mahavira >>  cpu:199.59% mem:0.01% io:0.00MB/s net:0.00MB/s (host:f307d28f60f4 image:agileek/cpuset-test)",

    "docker": {
        "image": "agileek/cpuset-test",
        "hostname": "f307d28f60f4",
        "created": "2016-12-21T15:04:27.6401807Z",
        "daemon_name": "jarvis",
        "name": "small_mahavira",
        "short_id": "f307d28f60",
        "status": "running",
        "pid": 7931,
        "id": "f307d28f60f4695e90bd108b3a3bbb6c9fbd4e896187966515b9d18d104f1730",
        "stats": {
          //raw stats and computed,from the docker daemon
        }
  }
}
```

# Incoming features

We are at early stage concerning this docker instrumentation any comments are welcome and we will do as much as we can to integrate the desired features!


CHANGELOG.md
# v1.2

Refactoring in order to use the official Docker Client Library (docker-py)
Features:

* Labels are now reported to each event in the field `custom.docker.labels` 
* New optsion `--matchByLabel` to filter container stats and logs based on labels
* Add a subsection `human_stats` to each "stats event" to provide more clarity


Human stats example:
```json
{
  "human_stats": {
    "cpu_stats": {
      "total_usage_%": 0,
      "usage_in_kernelmode_%": 0,
      "per_cpu_usage_%": [
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0
      ],
      "usage_in_usermode_%": 0
    },
    "networks": {
      "all": {
        "rx_bytes_ps": 0,
        "tx_bytes_ps": 0,
        "tx_packets_ps": 0,
        "rx_dropped_ps": 0,
        "tx_dropped_ps": 0,
        "tx_errors_ps": 0,
        "rx_errors_ps": 0,
        "rx_packets_ps": 0
      },
      "eth0": {
        "tx_packets_ps": 0,
        "rx_dropped_ps": 0,
        "rx_bytes_ps": 0,
        "tx_dropped_ps": 0,
        "tx_errors_ps": 0,
        "rx_errors_ps": 0,
        "tx_bytes_ps": 0,
        "rx_packets_ps": 0
      }
    },
    "memory_stats": {
      "usage_%": 0.0003174765431030924
    },
    "blkio_stats": {
      "read_bps": 0,
      "write_bps": 0,
      "total_bps": 0,
      "total_iops": 0,
      "write_iops": 0,
      "read_iops": 0
    }
  }
}
```
# Migrating from v1.0 to v1.2
The v2.0 is based on the official docker client python library.
If, you are currently using the v1.0, here are the breaking changes:

* Replace `--no-dockerEvents` option by `--no-events`
* Replace the `--dockerEvents` option by `--events`
* Replace `-h` by `--hostname`, `-h` is now displaying the help
* Replace `-a` by `--attr`, `-a` is no longer supported

The content of each event has also evolved.

The field `custom.docker.data_type` is no longer supported and is replaced 
by `custom.@marker`. Use this new field instead the previous one.



index.js
#! /usr/bin/env node

'use strict';

var net = require('net');
var eos = require('end-of-stream');
var through = require('through2');
var allContainers = require('docker-allcontainers');
var logFactory = require('docker-loghose');
var statsFactory = require('docker-stats');
var eventsFactory = require('docker-event-log');
var minimist = require('minimist');


function parseOptions(){
  var apiKey = process.argv[2];

  if(apiKey==null){
    console.log('You must define your Logmatic.io\'s api key as a first argument.\n' +
      '> logmatic-docker [apiKey] \n' +
                '   [-a ATTR (eg myattribute="my attribute")]\n' +
                '   [-h HOSTNAME (default "api.logmatic.io")] [-p PORT (default "10514")]\n' +
                '   [--namespace NAMESPACE (default "docker")]' +
                '   [--matchByImage REGEXP] [--matchByName REGEXP]\n' +
                '   [--skipByImage REGEXP] [--skipByName REGEXP]\n' +
                '   [--no-dockerEvents]\n' +
                '   [--no-logs]\n' +
                '   [--no-stats] [-i statsInterval]\n')

    process.exit(1);
  }

  var opts = minimist(process.argv.slice(3),{
              boolean: ["debug", "stats", "dockerEvents"],
              alias: {
                attr: "a",
                host: "h",
                port: "p",
                statsinterval: 'i'
              },
              default:{
                newline: true,
                stats: true,
                logs: true,
                namespace: "docker",
                dockerEvents: true,
                host: 'api.logmatic.io',
                port: '10514',
                apiKey: apiKey,
                statsinterval: 30
              }
            });

  //Parse extra attributes
  if(opts.attr){
    if(!Array.isArray(opts.attr)){
      opts.attr = [opts.attr];
    }
    var extra_attributes = {};
    for (var i in opts.attr) {
      var keyvalue = opts.attr[i].split('=');
      extra_attributes[keyvalue[0]] = keyvalue[1];
    }
    opts.attr = extra_attributes;
  }

  if(opts.debug){
    console.log("> Options:\n",opts);
  }
  return opts;
}

function merge(object, into) {
    if (!object) { return; }

    var key;
    for (key in object) {
      if (object.hasOwnProperty(key)) {
        into[key] = object[key];
    }
  }
}

function start() {
  var opts = parseOptions();

  var filter = through.obj(function(obj, enc, cb) {
    merge(opts.attr,obj);

    //Compute message
    var message;
    if (obj.line) {
      obj.data_type = "log";
      message = obj.line;
      delete obj.line;
    } else if (obj.type) {
      obj.data_type = "event";
      message = "[Docker event] host=\""+obj.host+"\" name=\""+obj.name+"\" event=\""+obj.type+"\"";
    } else if (obj.stats) {
      obj.data_type = "stats";

      //Compute mem pct
      obj.stats.memory_stats.memory_percent = obj.stats.memory_stats.usage/obj.stats.memory_stats.limit;

      message = "[Docker stats] host=\""+obj.image+"\" name=\""+obj.name+"\" main stats: [cpu%=" + Math.floor((obj.stats.cpu_stats.cpu_usage.cpu_percent*10000)/100)+ "% mem%="+Math.floor((obj.stats.memory_stats.memory_percent*10000)/100)+"%]";
    }

    //Handle namespacing
    if(opts.namespace){
        var newObj = {};
        newObj[opts.namespace] = obj;
        obj = newObj;
    }

    obj.message = message;

    if(opts.debug){
      console.log("> Send entry:\n",obj);
    }

    this.push(opts.apiKey);
    this.push(' ');
    this.push(JSON.stringify(obj));
    this.push('\n');

    cb()
  });

  var events = allContainers(opts);
  var streamsOpened = 0;

  if (opts.logs) {
      var loghose = logFactory(opts);
      loghose.pipe(filter);
      streamsOpened++;
  }

  if (opts.stats) {
      var stats = statsFactory(opts);
      stats.pipe(filter);
      streamsOpened++;
  }


  if (opts.dockerEvents) {
      var dockerEvents = eventsFactory(opts);
      dockerEvents.pipe(filter);
      streamsOpened++;
  }

  pipe();

  // destroy out if all streams are destroyed
  loghose && eos(loghose, function() {
    streamsOpened--;
    streamClosed(streamsOpened);
  });
  dockerEvents && eos(dockerEvents, function() {
    streamsOpened--;
    streamClosed(streamsOpened);
  });

  return loghose;

  var out;
  var noRestart = function() {};

  function pipe() {
    if (out) {
      filter.unpipe(out);
    }

    out = net.createConnection(opts.port, opts.host);
    filter.pipe(out, { end: false });
    noRestart = eos(out, pipe);
  }

  function streamClosed(streamsOpened) {
    if (streamsOpened <= 0) {
      noRestart()
      out.destroy();
    }
  }
}

module.exports = start;

if (require.main === module) {
  start();
}


agent/Calculator.py
import logging
import time

logger = logging.getLogger()


class Calculator:
    def __init__(self):
        self.datastore = {}

    def _delta_meter_ps(self, key, new_tick, new_value):
        delta = 0.0
        if key in self.datastore:
            delta = new_value - self.datastore[key]["value"]
            delta = delta / (new_tick - self.datastore[key]["tick"])

        self.datastore[key] = {"tick": new_tick, "value": new_value, "delta": delta}
        return delta

    def compute_human_stats(self, container, stats, detailed):

        # Add computed stats

        tick = int(time.time() * 1000)

        computed_stats = {
            "blkio_stats": {},
            "memory_stats": {},
            "cpu_stats": {},
            "networks": {}
        }
        if detailed is True:
            computed_stats.update(stats)

        # enrich stats
        computed_stats["blkio_stats"].update(self._compute_blkio(tick, container, stats))
        computed_stats["memory_stats"].update(self._compute_memory(stats))
        computed_stats["cpu_stats"].update(self._compute_cpu(stats))

        network_stats = self._compute_network(tick, container, stats)
        for interface in network_stats:
            if interface not in computed_stats["networks"]:
                computed_stats["networks"][interface] = {}
            computed_stats["networks"][interface].update(network_stats[interface])

        return computed_stats

    def _compute_cpu(self, stats):
        try:

            old = stats["precpu_stats"]["cpu_usage"]
            new = stats["cpu_stats"]["cpu_usage"]

            per_cpu = []
            for i in range(len(new["percpu_usage"])):
                per_cpu.append((new["percpu_usage"][i] - old["percpu_usage"][i]) / 1000000000)

            total = (new["total_usage"] - old["total_usage"]) / 1000000000
            user = (new["usage_in_usermode"] - old["usage_in_usermode"]) / 1000000000
            kernel = (new["usage_in_kernelmode"] - old["usage_in_kernelmode"]) / 1000000000

            return {
                "per_cpu_usage_pct": per_cpu,
                "total_usage_pct": total,
                "usage_in_kernelmode_pct": kernel,
                "usage_in_usermode_pct": user
            }
        except Exception as e:
            return {"error": {"message": "Couldn't compute CPU stats (API Version): " + str(e)}}

    def _compute_memory(self, stats):
        try:
            return {
                "usage_pct": stats["memory_stats"]["usage"] / stats["memory_stats"]["limit"],
            }
        except Exception as e:
            return {"error": {"message": "Couldn't compute memory stats (API Version): " + str(e)}}

    def _compute_blkio(self, tick, c, stats):

        try:
            summed = {
                "io": {"Read": 0, "Write": 0, "Total": 0},
                "bs": {"Read": 0, "Write": 0, "Total": 0}
            }

            # compute all IO data (bytes per second) into write/read/total
            for entry in stats["blkio_stats"]["io_service_bytes_recursive"]:
                if entry["op"] in summed["bs"]:
                    summed["bs"][entry["op"]] = summed["bs"][entry["op"]] + entry["value"]
            # compute all IO data (io per second) into write/read/total
            for entry in stats["blkio_stats"]["io_serviced_recursive"]:
                if entry["op"] in summed["io"]:
                    summed["io"][entry["op"]] = summed["io"][entry["op"]] + entry["value"]

            return {
                "read_bps": self._delta_meter_ps(c.short_id + ".blk.read", tick, summed["bs"]["Read"]),
                "write_bps": self._delta_meter_ps(c.short_id + ".blk.write", tick, summed["bs"]["Write"]),
                "total_bps": self._delta_meter_ps(c.short_id + ".blk.total", tick, summed["bs"]["Total"]),
                "read_iops": self._delta_meter_ps(c.short_id + ".blk.io.read", tick, summed["io"]["Read"]),
                "write_iops": self._delta_meter_ps(c.short_id + ".blk.io.write", tick, summed["io"]["Write"]),
                "total_iops": self._delta_meter_ps(c.short_id + ".blk.io.total", tick, summed["io"]["Total"])
            }
        except Exception as e:
            return {"error": {"message": "Couldn't compute BLKIO stats (API Version): " + str(e)}}

    def _compute_network(self, tick, c, stats):
        try:
            network = {}

            nets = stats["networks"]
            network["all"] = {}

            for interface in nets:
                network[interface] = {}
                for metric in nets[interface]:
                    key = "{}.{}.{}".format(c.short_id, interface, metric)
                    value = nets[interface][metric]
                    if metric + "_ps" not in network["all"]:
                        network["all"][metric + "_ps"] = 0
                    network["all"][metric + "_ps"] += value
                    network[interface][metric + "_ps"] = self._delta_meter_ps(key, tick, value)

                for metric in network["all"]:
                    key = "{}.net.{}".format(c.short_id, metric)
                    value = network["all"][metric]
                    network["all"][metric] = self._delta_meter_ps(key, tick, value)

            return network

        except Exception as e:
            return {"error": {"message": "Couldn't compute networks stats (API Version): " + str(e)}}


agent/AgentReporter.py
import logging
import re

from agent.Calculator import Calculator

logger = logging.getLogger()


class AgentReporter:
    def __init__(self, client, logger, args):
        self.client = client
        self.logger = logger
        self.args = args
        self.calculator = Calculator()
        self.attrs = {}
        self.daemon_name = self.client.info()["Name"]
        self.local_cache = {}

        for attr in args.attrs:
            kv = attr.split("=")
            if len(kv) > 1:
                self.attrs[kv[0]] = kv[1]
            else:
                self.attrs[kv[0]] = ""

    def export_events(self):
        """Send events to Logmatic.io"""
        try:
            events = self.client.events(decode=True)
            for event in events:
                if event["Type"] == "container":
                    meta = self._build_meta_from_event(event)
                    meta["@marker"] = ["docker", "docker-events"]
                    meta[self.args.ns]["event"] = event["Action"]
                    meta[self.args.ns]["status"] = event["status"]
                    if "exitCode" in event["Actor"]["Attributes"]:
                      meta[self.args.ns]["exit_code"] = event["Actor"]["Attributes"]["exitCode"]

                    # send it to Logmatic.io
                    self.logger.info("[Docker event] name:{} >> event:{} (image={})"
                                     .format(event["Actor"]["Attributes"]["name"],
                                             event["Action"],
                                             event["Actor"]["Attributes"]["image"]),
                                     extra=meta)

        except Exception as e:
            logger.exception("Unexpected error during the processing of events: {}".format(e))

    def export_stats(self, container, detailed):
        """Send container stats to Logmatic.io"""
        try:
            meta = self._build_context(container)
            meta["@marker"] = ["docker", "docker-stats"]

            # call the API
            stats = container.stats(stream=False, decode=True)
            computed_stats = self.calculator.compute_human_stats(container, stats, detailed)
            meta[self.args.ns]["stats"] = computed_stats

            # format the event message
            message = ""
            if "error" not in computed_stats["cpu_stats"]:
                message += " cpu:{:.2f}%".format(computed_stats["cpu_stats"]["total_usage_pct"] * 100.0)
            if "error" not in computed_stats["memory_stats"]:
                message += " mem:{:.2f}%".format(computed_stats["memory_stats"]["usage_pct"] * 100.0)
            if "error" not in computed_stats["blkio_stats"]:
                message += " io:{:.2f}MB/s".format(computed_stats["blkio_stats"]["total_bps"] / 1000000.0)
            if "error" not in computed_stats["networks"]:
                message += " net:{:.2f}MB/s".format(
                    (computed_stats["networks"]["all"]["tx_bytes_ps"] + computed_stats["networks"]["all"][
                        "rx_bytes_ps"]) / 1000000.0)

            # send to Logmatic.io
            self.logger.info("[Docker stats] name:{} >> {} (host:{} image:{})"
                             .format(meta[self.args.ns]["name"],
                                     message,
                                     meta[self.args.ns]["hostname"],
                                     meta[self.args.ns]["image"]),
                             extra=meta)

        except Exception as e:
            logger.exception("Unexpected error during the processing of stats: {}".format(e))

    def export_logs(self, container):
        try:
            line = ""
            meta = self._build_context(container)
            meta["@marker"] = ["docker", "docker-logs"]
            logs = container.logs(stream=True, stderr=True, stdout=True, follow=True, tail=0)
            for chunk in logs:

                # Append all char into a string until a \n
                if type(chunk) is not str:
                    logger.debug("Decode needed:{}".format(chunk))
                    chunk = chunk.decode()

                line = line + chunk
                if line.endswith("\n"):
                    logger.debug("Concat done: {}".format(line))
                    self.logger.info(line, extra=meta)
                    line = ""


        except Exception as e:
            logger.exception("Unexpected error during the processing of logs: {}".format(e))

    def _build_context(self, container):
        """Internal method, build the container context"""
        try:

            # Concatenate all labels
            labels = {}
            if len(container.attrs["Config"]["Labels"]) > 0:
                labels["all"] = []
                for label in container.attrs["Config"]["Labels"]:
                    labels["all"].append(label)
                    if container.attrs["Config"]["Labels"][label] != "":
                        labels[label] = container.attrs["Config"]["Labels"][label]

            # Add all container/image information
            meta = {
                self.args.ns: {
                    "id": container.id,
                    "short_id": container.short_id,
                    "name": container.name,
                    "status": container.status,
                    "daemon_name": self.daemon_name,
                    "labels": labels,
                    "hostname": container.attrs["Config"]["Hostname"],
                    "image": container.attrs["Config"]["Image"],
                    "created": container.attrs["Created"],
                    "pid": container.attrs["State"]["Pid"]
                },
                "severity": "INFO"
            }

            # Add all attributes
            if len(self.attrs):
                meta["attr"] = self.attrs

            return meta.copy()

        except Exception as e:
            logger.exception("Unexpected error during the processing of context: {}".format(e))

    def filter(self, containers):
        """Expose only the containers and the images that match the rules set
            - skip_image: exclude all images matching the regex
            - match_image: keep all images matching the regex
            - skip_container: exclude all containers matching the regex
            - match_container: keep all containers matching the regex
        """
        filtered = []

        for c in containers:
            if c.attrs["Config"]["Image"].startswith("logmatic/logmatic-docker"):
                continue
            # skip container by image name
            if self.args.skip_image and re.search(self.args.skip_image, c.attrs["Config"]["Image"]):
                continue
            # skip container by container name
            if self.args.skip_name and re.search(self.args.skip_name, c.name):
                continue
            # match container by image name
            if self.args.match_image and not re.search(self.args.match_image, c.attrs["Config"]["Image"]):
                continue
            # match container by name name
            if self.args.match_name and not re.search(self.args.match_name, c.name):
                continue

            # build local meta for each container
            self.local_cache[c.id] = self._build_context(c)

            filtered.append(c)

        logger.debug("Filter // Keeping containers: {}".format(filtered))
        return filtered

    def _build_meta_from_event(self, event):

        if event["id"] in self.local_cache:
            return self.local_cache[event["id"]].copy()

        try:
            container = self.client.containers.get(event["id"])
            meta = self._build_context(container)
            return meta
        except:
            logger.exception("Could not find container {}".format(event["id"]))
            # fallback

            # Add all container/image information
            meta = {
                self.args.ns: {
                    "id": event["id"],
                    "short_id": event["id"][:12],
                    "name": event["Actor"]["Attributes"]["name"],
                    "daemon_name": self.daemon_name,
                    "image": event["Actor"]["Attributes"]["image"]
                },
                "severity": "INFO"
            }
            return meta


main.py
#!/usr/bin/python
import argparse
import logging
import sys
import threading
from time import sleep

import docker
import logmatic

from agent.AgentReporter import AgentReporter


internal_logger = logging.getLogger()

# Args parser settings
parser = argparse.ArgumentParser(prog="logmatic-docker", description='Send logs, events and stats to Logmatic.io')
parser.add_argument("token", metavar="LOGMATIC_API_KEY", help='The Logmatic.io API key')
parser.add_argument("--no-ssl", dest='ssl', action="store_false", help="Do not use ssl connection")
parser.add_argument("--logs", dest='logs', action="store_true", help="Enable the logs streams")
parser.add_argument('--no-logs', dest='logs', action="store_false", help="Disable the logs streams")
parser.add_argument("--stats", dest='stats', action="store_true", help="Enable the stats streams")
parser.add_argument('--no-stats', dest='stats', action="store_false", help="Disable the stats streams")
parser.add_argument('--no-detailed-stats', dest='detailed_stats', action="store_false", help="Disable stats streams")
parser.add_argument("--events", dest='events', action="store_true", help="Enable the event stream")
parser.add_argument('--no-events', dest='events', action="store_false", help="Disable the event stream")
parser.add_argument("--namespace", dest='ns', help="Default namespace")
parser.add_argument("--hostname", dest='hostname', help="Logmatic.io's hostname (default api.logmatic.io)")
parser.add_argument("--port", dest='port', type=int, help="Logmatic.io's port (default 10514)")
parser.add_argument("--debug", dest="debug", action="store_true", help="Enable debugging")
parser.add_argument("-i", dest='interval', type=int, help="Seconds between to stats report (default 30)")
parser.add_argument("--attr", dest="attrs", action='append', help="eg my_attribute=\"my attribute\"")
parser.add_argument("--docker-version", dest='docker_version', metavar="VER", help="Force the Docker version to use")
parser.add_argument("--skipByImage", dest='skip_image', metavar="REGEX", help="Skip container by image name")
parser.add_argument("--skipByName", dest='skip_name', metavar="REGEX", help="Skip container by container name")
parser.add_argument("--matchByImage", dest='match_image', metavar="REGEX", help="Match container by image name")
parser.add_argument("--matchByName", dest='match_name', metavar="REGEX", help="Match container by container name")
parser.add_argument("--matchByLabel", dest='match_label', metavar="LABEL", help="Format either \"key\" or \"key=value\"")

# Default values
parser.set_defaults(logs=True)
parser.set_defaults(stats=True)
parser.set_defaults(detailed_stats=True)
parser.set_defaults(events=True)
parser.set_defaults(ns="docker")
parser.set_defaults(hostname="api.logmatic.io")
parser.set_defaults(port=10515)
parser.set_defaults(ssl=True)
parser.set_defaults(interval=30)
parser.set_defaults(attrs=[])
parser.set_defaults(debug=False)
parser.set_defaults(docker_version="auto")
parser.set_defaults(skip_name=None)
parser.set_defaults(skip_image=None)
parser.set_defaults(match_name=None)
parser.set_defaults(match_image=None)
parser.set_defaults(match_label=None)

args = parser.parse_args()

# Initialise the logger for Logmatic.io
logmatic_logger = logging.getLogger("docker-logmatic")
handler = logmatic.LogmaticHandler(args.token, host=args.hostname, port=args.port, ssl=args.ssl)
handler.setFormatter(logmatic.JsonFormatter(fmt="%(message)"))
logmatic_logger.addHandler(handler)
logmatic_logger.setLevel(logging.DEBUG)
logmatic_logger.propagate = False

if args.debug is True:
    internal_logger.setLevel(logging.DEBUG)
    sys_handler = logging.StreamHandler(sys.stderr)
    internal_logger.addHandler(sys_handler)
    internal_logger.debug(args)
else:
    internal_logger.disabled = True


# Initialise the connection to the local daemon
base_url = 'unix://var/run/docker.sock'
client = docker.DockerClient(base_url=base_url, timeout=None, version=args.docker_version)

# Main logic starts here
agent = AgentReporter(client=client, logger=logmatic_logger, args=args)
filters = None
if args.match_label:
    filters = {"label": args.match_label}

# Initialize all threads
event_thread = None
log_threads = {}
logs = False

# Main loop
while 1:

    try:
        containers = None
        # Start the event thread, and check if it's alive each seconds
        if args.events is True and (event_thread is None or not event_thread.isAlive()):
            internal_logger.info("Starting the event stream thread")
            event_thread = threading.Thread(target=agent.export_events)
            event_thread.daemon = True
            event_thread.start()

        # Start all log threads, and check if they're alive each x seconds
        if args.logs is True or args.stats is True:
            containers = client.containers.list(filters=filters)
            containers_filtered = agent.filter(containers)
            for container in containers_filtered:
                # Start threads and check if each logging thread are alive
                if args.logs is True and (container.id not in log_threads or not log_threads[container.id].isAlive()):
                    internal_logger.info("Starting the log stream thread for " + container.id)
                    log_threads[container.id] = threading.Thread(target=agent.export_logs, args=[container])
                    log_threads[container.id].daemon = True
                    log_threads[container.id].start()

                # Export stats to Logmatic.io
                if args.stats is True:
                    agent.export_stats(container, detailed=args.detailed_stats)

        internal_logger.debug("Next tick in {}s".format(args.interval))
        sleep(args.interval)

    except (KeyboardInterrupt, SystemExit):
        exit(0)
